internal/diff/diff.go

```diff
@@ -0,0 +1,153 @@
+package diff
+
+import (
+	"cmp"
+	"slices"
+)
+
+type diffResult struct {
+	Size     int64         `json:"size"`
+	Packages []diffPackage `json:"packages"`
+	Sections []diffSection `json:"sections"`
+}
+
+type changeType string
+
+const (
+	changeTypeAdd    changeType = "add"
+	changeTypeRemove changeType = "remove"
+	changeTypeChange changeType = "change"
+)
+
+type diffBase struct {
+	Name       string     `json:"name"`
+	From       int64      `json:"from"`
+	To         int64      `json:"to"`
+	ChangeType changeType `json:"change_type"`
+}
+
+func diffBaseCmp(a, b diffBase) int {
+	return -cmp.Compare(a.To-a.From, b.To-b.From)
+}
+
+type diffPackage struct {
+	diffBase
+}
+
+type diffSection struct {
+	diffBase
+	oldFileSize  int64
+	oldKnownSize int64
+	newFileSize  int64
+	newKnownSize int64
+}
+
+func newDiffResult(newResult, oldResult *commonResult) diffResult {
+	ret := diffResult{
+		Packages: make([]diffPackage, 0),
+		Sections: make([]diffSection, 0),
+	}
+
+	// diff packages
+	for k, v := range newResult.Packages {
```

Comment:
The complexity of this for + if +( if )+ else is insane.

You could simplify by inverting boolean check you do and use continue at least at 2 places.

I'm on my phone I cannot help but I think you should